alternator/ttl.cc

```diff
@@ -113,4 +139,321 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+            // FIXME: need to split the scan up into ranges owned by this node
+            // *AND* this shard - instead of scanning the entire range. Note
+            // because we're doing a coordinator scan, we need to limit
+            // ourselves to the ranges of this shard - it doesn't happen
+            // automatically.
+            // FIXME: consider if we should ask the scan without caching?
+            // can we use cache but not fill it?
+            // FIXME: don't read the entire items - read only parts of it.
+            // We must read the key columns (to be able to delete) and also
+            // the requested attribute. If the requested attribute is a map's
+            // member we may be forced to read the entire map - but it would
+            // be good if we can read only the single item of the map - it
+            // should be possible (and a must for issue #7751!).
+            lw_shared_ptr<service::pager::paging_state> paging_state = nullptr;
+            auto regular_columns = boost::copy_range<query::column_id_vector>(
+            s->regular_columns() | boost::adaptors::transformed([] (const column_definition& cdef) { return cdef.id; }));
+            auto selection = cql3::selection::selection::wildcard(s);
+            query::partition_slice::option_set opts = selection->get_query_options();
+            opts.set<query::partition_slice::option::allow_short_read>();
+            std::vector<query::clustering_range> ck_bounds{query::clustering_range::make_open_ended_both_sides()};
+            auto partition_slice = query::partition_slice(std::move(ck_bounds), {}, std::move(regular_columns), opts);
+            auto command = ::make_lw_shared<query::read_command>(s->id(), s->version(), partition_slice, _proxy.get_max_result_size(partition_slice));
+            executor::client_state client_state{executor::client_state::internal_tag()};
+            tracing::trace_state_ptr trace_state;
+            // FIXME: is empty_service_permit() the right thing?
+            // view builder has _permit = _db.get_reader_concurrency_semaphore().make_tracking_only_permit(nullptr, "view_builder", db::no_timeout))
+            auto query_state_ptr = std::make_unique<service::query_state>(client_state, trace_state, empty_service_permit());
+            // FIXME: What should we do on multi-DC? Will we run the expiration on the same ranges on all
+            // DCs or only once for each range? If the latter, we need to change the CLs in the
+            // scanner and deleter.
+            db::consistency_level cl = db::consistency_level::LOCAL_QUORUM;
+            auto query_options = std::make_unique<cql3::query_options>(cl, std::vector<cql3::raw_value>{});
+            query_options = std::make_unique<cql3::query_options>(std::move(query_options), std::move(paging_state));
+            dht::partition_range_vector partition_ranges;
+            partition_ranges.push_back(dht::partition_range::make_open_ended_both_sides());
```

Comment:
This should be limited to primary owned ranges, to make sure only one node at a time tries to expire a particular row, and to this shard's ranges.

Reply:
Yes, this is also in a FIXME above.

Reply:
Was this FIXME fixed?

Reply:
Yes, in patch 12.. It looks like you want me to conflate the entire scanner implementation into one patch instead of introducing a simpler one (scanning everything) in this patch and then fixing it to scan only the data owned by each shard in patch 12. Ok then, I'll merge the two patches.

Reply:
The two patches were merged.